### PR TITLE
Archive inboxes after 5 days of delivery failures 

### PR DESCRIPTION
**Pull request by annando**
Currently we keep on delivering AP contacts, even when they aren't reachable for months. Due to the nature of the AP network, we cannot use the "archive" field in the contact.

So we now do have a new table. This stores the status of the inbox. And it prevents delivering content when the inbox hadn't been accessible for some time.

### CODE REVIEW

File: config/dbstructure.config.php [hunk 1]
```diff
@@ -529,6 +529,21 @@
...
 	],
+	"inbox-status" => [
+		"comment" => "Status of ActivityPub inboxes",
+		"fields" => [
+			"url" => ["type" => "varbinary(255)", "not null" => "1", "primary" => "1", "comment" => "URL of the inbox"],
+			"created" => ["type" => "datetime", "not null" => "1", "default" => DBA::NULL_DATETIME, "comment" => "Creation date of this entry"],
+			"success" => ["type" => "datetime", "not null" => "1", "default" => DBA::NULL_DATETIME, "comment" => "Date of the last successful delivery"],
+			"failure" => ["type" => "datetime", "not null" => "1", "default" => DBA::NULL_DATETIME, "comment" => "Date of the last failed delivery"],
+			"previous" => ["type" => "datetime", "not null" => "1", "default" => DBA::NULL_DATETIME, "comment" => "Previous delivery date"],
+			"archive" => ["type" => "boolean", "not null" => "1", "default" => "0", "comment" => "Is the inbox archived?"],
```
Comment (MrPetovan) [inline config/dbstructure.config.php]: Please rename this field to `archived`.

File: database.sql [hunk 2]
```diff
@@ -470,6 +470,20 @@ CREATE TABLE IF NOT EXISTS `hook` (
...
+--
+-- TABLE inbox-status
+--
+CREATE TABLE IF NOT EXISTS `inbox-status` (
+	`url` varbinary(255) NOT NULL COMMENT 'URL of the inbox',
+	`created` datetime NOT NULL DEFAULT '0001-01-01 00:00:00' COMMENT 'Creation date of this entry',
+	`success` datetime NOT NULL DEFAULT '0001-01-01 00:00:00' COMMENT 'Date of the last successful delivery',
+	`failure` datetime NOT NULL DEFAULT '0001-01-01 00:00:00' COMMENT 'Date of the last failed delivery',
+	`previous` datetime NOT NULL DEFAULT '0001-01-01 00:00:00' COMMENT 'Previous delivery date',
+	`archive` boolean NOT NULL DEFAULT '0' COMMENT 'Is the inbox archived?',
```
Comment (MrPetovan) [inline database.sql]: Please rename this field to `archived`.

File: src/Model/APContact.php [hunk 3]
```diff
@@ -231,4 +234,27 @@ public static function getByURL($url, $update = null)
...
+	 */
+	private static function unarchiveInbox($url, $shared)
+	{
+		if (empty($url)) {
+			return;
+		}
+
+		$now = DateTimeFormat::utcNow();
+
+		$fields = ['archive' => false, 'success' => $now, 'shared' => $shared];
```
Comment (MrPetovan) [inline src/Model/APContact.php]: Please rename this field to `archived`.

File: src/Protocol/ActivityPub/Transmitter.php [hunk 4]
```diff
@@ -465,6 +465,18 @@ private static function createPermissionBlockForItem($item, $blindcopy, $last_id
...
+	/**
+	 * Check if an inbox is archived
+	 *
+	 * @param string $url Inbox url
+	 *
+	 * @return boolean "true" if inbox is archived
+	 */
+	private static function archivedInbox($url)
+	{
+		return DBA::exists('inbox-status', ['url' => $url, 'archive' => true]);
```
Comment (MrPetovan) [inline src/Protocol/ActivityPub/Transmitter.php]: Please rename this field to `archived`.

File: src/Util/HTTPSignature.php [hunk 5]
```diff
@@ -314,7 +315,66 @@ public static function transmit($data, $target, $uid)
...
+			$stamp2 = strtotime($now);
+			$previous_stamp = strtotime($status['previous']);
+
+			// Archive the inbox when there had been failures for five days.
+			// Additionally ensure that at least one previous attempt has to be in between.
+			if ((($stamp2 - $stamp1) >= 86400 * 5) && ($previous_stamp > $stamp1)) {
+				$fields['archive'] = true;
+			}
+		} else {
+			$fields['archive'] = false;
```
Comment (MrPetovan) [inline src/Util/HTTPSignature.php]: Please rename this field to `archived`.

File: src/Util/HTTPSignature.php [hunk 6]
```diff
@@ -314,7 +315,66 @@ public static function transmit($data, $target, $uid)
...
+				$stamp1 = strtotime($status['success']);
+			}
+
+			$stamp2 = strtotime($now);
+			$previous_stamp = strtotime($status['previous']);
+
+			// Archive the inbox when there had been failures for five days.
+			// Additionally ensure that at least one previous attempt has to be in between.
+			if ((($stamp2 - $stamp1) >= 86400 * 5) && ($previous_stamp > $stamp1)) {
+				$fields['archive'] = true;
```
Comment (MrPetovan) [inline src/Util/HTTPSignature.php]: Please rename this field to `archived`.

**Comment by annando** [comment]
I had named the "archive" field after the similar field in the contact table.

**Comment by MrPetovan** [comment]
And it was wrongly named too.

**Comment by annando** [comment]
Sorry. I mostly orientate myself after existing stuff. Sadly this is now merged, so changing this would be more work.

**Comment by MrPetovan** [comment]
Yeah, I know, but I thought I would point is out. Boolean fields are (or should be) named with a past participle to indicate it's a boolean flag without having to dig into the table structure.

**Pull request merged**